### ISSUE TITLE
Fuzzy sarch

### ISSUE THREAD

**Issue by dreua** [osm-search/Nominatim#2306]
**Is your feature request related to a problem? Please describe.**
As a user I often don't know the exact name of what I'm searching for and I don't get results if they are not exactly in the OSM tags. 
Examples:
- Searching for "tinten drechsler mainz" I get no results, but "tintencenter drechsler mainz" or just "drechsler mainz" yields exactly what I'm looking for.
- Searching for "hallenbad mombach" yields not results, but "hallen mombach" does, becaus it is tagged as "Hallen- und Freibad Mombach". (Don't think we should tag for the map or the search algorithm here.)

**Describe the solution you'd like**
I'd like to get results of entities containing tags with similar words, especially in cases where the result set would otherwise be empty.

**Describe alternatives you've considered**
I really don't want to suggest Google or Apple maps here ;)

**Additional context**

I think the search terms above should be sufficient, let me know if you need more examples, I stumble upon them every now end then. Some fuzzy matching to allow for typos would be great, but let's start with fixing these examples here.

**Comment by RhinoDevel** [comment]
Maybe [Photon](https://github.com/komoot/photon) is a solution for you?

**Comment by lonvia** [comment]
For the second case you probably want to add 'Hallenbad' as a special phrase in https://wiki.openstreetmap.org/wiki/Nominatim/Special_Phrases/DE. The search is rather "get me some swimming pool in Mombach" rather than "get me the thing named Hallenbad in Mombach". That's what special phrases are for.

The first case is not quite as clear cut. You seem to be searching for an ink supplier called 'Drechsler" in Mainz. So that's the whole category of questions like 'where can I get ink in Mainz?' and that is slightly out of scope for Nominatim.

**Comment by dreua** [comment]
Thanks for the link to "Special prases"! I'll edit that accordingly. Will changes to this page automatically be imported to Nominatim (if yes, how fast)?

I tried Photon but it didn't give me relevant results in those two cases.

**Comment by dreua** [comment]
> The first case is not quite as clear cut. You seem to be searching for an ink supplier called 'Drechsler" in Mainz. So that's the whole category of questions like 'where can I get ink in Mainz?' and that is slightly out of scope for Nominatim.

No, I was indeed searching for a that specific shop, I just wasn't 100% sure about the name: I knew that it contained the words "tinten" and "drechsler" but wasn't sure whether it is "tintendrechsler" or "tinten- und toneladen drechsler" or "tintencenter drechsler". I'd be happy if nominatim could match the word "tinten" as an exact substring of "tintencenter". This may not be as important in english, but it is for languages with compound words like German.

**Comment by lonvia** [comment]
We'd have to do compound decomposition here. It might be a possibility in the future, especially now that we have the possibility to implement language-specific tokenizers. But it's still a bit off.

**Comment by lonvia** [comment]
A quick search yields
* https://github.com/repodiac/german_compound_splitter
* and https://github.com/JoelNiklaus/CompoundSplit

They look like a good base for a custom German tokenizer.
